Feature/issue#12 - Spotify 소셜 로그인 구현

minju412 opened the pull request:
### ✍️ Description
- OAuth 2.0 을 사용해 소셜 로그인을 구현하였으며, Provider는 Spotify 입니다.

<br>

### 🎯 Key Changes
- `oauth.handler.CustomOAuth2AuthenticationSuccessHandler`: 스포티파이 로그인 성공 시, 사용자가 요청했던 `originalUri`로 리다이렉트합니다.
- `oauth.handler.OAuth2AuthenticationFailureHandler`: 로그인 실패 시 로그를 남기고 Exception을 발생시킵니다.
- `oauth.provider.OAuth2Provider`: Provider의 종류와 정보를 나타내는 Enum으로 현재는 `SPOTIFY`만 포함되어 있습니다.
- `oauth.provider.OAuth2UserInfo`: OAuth2User 정보를 나타내는 클래스입니다.
- `oauth.provider.SpotifyUserInfo`: Spotify 사용자 정보를 나타내는 클래스입니다.
- `oauth.CustomOAuth2User`: `DefaultOAuth2User`를 상속한 클래스로 Provider와 엑세스 토큰을 가지고 있습니다.
- `oauth.CustomOAuth2UserService`: `code`를 통해 사용자 정보를 로드하고 데이터베이스에 저장합니다.

<br>

### 💬 To Reviewers
- 자체 JWT는 발급할 필요가 없을 것 같다고 판단으로 JWT 발급 로직은 제거했습니다.
    - 저희 앱을 사용하는 유저가 모두 스포티파이 유저라고 가정했을 때, 스포티파이 인증 서버를 통해 생성한 인증 정보(authentication)을 활용하면 된다고 생각했기 때문입니다!

minju412 committed:
chore: 인증 관련 의존성 추가

- Spotify 로그인을 이용하기 위해 OAuth2-Client 사용

minju412 committed:
feat: Spotify 소셜 로그인 구현

- 허용되지 않은 uri에 대해 스포티파이 로그인으로 리다이렉트
- 스포티파이 로그인 성공 후 사용자의 이름, 이메일, 프로필 이미지 정보를 데이터베이스에 저장
- 로그인 성공 후 스포티파이에서 내려준 엑세스 토큰을 쿼리에 담아 응답

minju412 committed:
fix: Resolve Conflicts

minju412 committed:
Merge branch 'develop' into feature/issue#12

Jeonwoojae commented on src/main/java/com/petitapetit/miml/domain/auth/oauth/CustomOauth2UserService.java:
```diff
@@ -0,0 +1,74 @@
+			isFirst = false;
+			member = userOptional.get();
+			// member.updateEmail(oAuth2UserInfo.getEmail());
+			memberRepository.save(member);
```
해당 부분에서 단순히 db에 저장된 내용을 가져와서 다시 저장하고 있습니다
업데이트 부분이 주석 되어있는데 의도한 것인가요?🤔

minju412 replied:
은서님께서도 같은 부분에 코멘트를 남겨주셨는데요, 해당 로직은 불필요한 로직이라고 판단되어 삭제했습니다!

Jeonwoojae commented on src/main/java/com/petitapetit/miml/domain/member/repository/MemberRepository.java:
```diff
@@ -0,0 +1,19 @@
+import com.petitapetit.miml.domain.auth.oauth.provider.OAuth2Provider;
+
+@Repository
+@Transactional
```
`@Transactional`은 보통 서비스 레이어에서 많이 사용하는데 레포지토리에서 사용하시게 된 이유가 궁금합니다!

minju412 replied:
습관적으로 사용했던 것 같은데, 고민해보니 레포지토리에서는 트랜잭션을 사용할 필요가 없을 것 같네요..! 제거하겠습니다! 👍

minju412 committed:
Merge branch 'develop' into feature/issue#12

kdmstj commented on build.gradle:
gson 라이브러리가 중복으로 추가 되었습니다.
```
implementation 'com.google.code.gson:gson'
```

minju412 replied:
앗 그렇네요..! 🤭 제거하겠습니다!

kdmstj commented on src/main/java/com/petitapetit/miml/domain/auth/oauth/CustomOauth2UserService.java:
1. isFirst 변수로 최초 로그인 여부를 구분한 이유가 있을까요? Optional을 통해 최초 로그인 여부를 확인할 수 있다고 생각되는데 변수를 설정하신 이유가 궁금합니다!

2. 더불어 이미 로그인한 적이 있는 유저에 대한 정보를 다시 DB에 저장한 이유에 대해 궁금합니다.

3. processOAuth2User 메소드에 여러개의 로직(Oauth2 서비스 제공자 구분, 새로운 사용자 DB 저장, OAuth2 User 생성)이 포함되어있다고 생각되는데 분리하는 것은 어떨까요?

minju412 replied:
1. 말씀하신대로 Optional을 이용하여 isEmpty일 경우 최초 로그인한 사용자겠네요..! isFirst 변수 삭제했습니다!
2. 요것도 불필요한 로직이라고 생각되어 제거했습니다.
3. 클린 코드에 대한 아이디어 너무 감사합니다! 한 가지 역할만 하도록 분리했는데, 코드로 확인 부탁드립니다! 👍

kdmstj commented on src/main/java/com/petitapetit/miml/domain/auth/oauth/api/LoginController.java:
1. @RestController 어노테이션을 작성하신 이유가 궁금합니다
해당 어노테이션은 @Controller에 @ResponseBody가 추가된 것으로 주 용도로는 JSON 형태로 객체 데이터를 반환하는 것으로 알고 있습니다. 단순 View를 반환하고자 하면 @Controller 어노테이션을 사용하는 것은 어떨까요?

minju412 replied:
습관적으로 `@RestController`를 사용했던 것 같은데, 맞는 말이네요! 현재 코드에서 LoginController가 제거되어서 적용할 수는 없지만 유의하겠습니다.

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/member/model/Member.java:
```diff
@@ -1,27 +1,41 @@
+	@Enumerated(EnumType.STRING)
+	private RoleType role;
+
+	@Enumerated(EnumType.STRING)
+	private OAuth2Provider provider;
+
+	private String providerId;
```
`NIT`: 해당 정보들은 따로 묶는것도 좋을 것 같습니다!
```suggestion
	@Enumerated(EnumType.STRING)
	private OAuthInfo oAuthInfo;
	
public class OAuthInfo {
    private OAuth2Provider provider;
    private String providerId;
}
```

minju412 replied:
좋습니다!! 해당 코멘트에서 인사이트를 얻어서 복합키를 적용해 보았는데요, `provider`와 `providerId`를 묶어 사용자를 식별하는데 사용합니다. 좋은 리뷰 감사합니다! 👍

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/member/repository/MemberRepository.java:
```diff
@@ -0,0 +1,19 @@
+@Repository
+@Transactional
+public interface MemberRepository extends JpaRepository<Member, Long> {
+	Optional<Member> findByProviderId(String providerId);
```
음.. 아래의 `findByProviderId()`와 같은 기능을 할 것 같은데 조금 위험해 보이네요.
providerId 동일한 케이스는 정말 희박하겠지만 그래도 `OAuth2Provider` 까지 비교하는 방어적인 코드를 짜는게 좋다고 생각해요!

minju412 replied:
맞습니다! `findByProviderId` 메서드는 코드 내부에서 사용되고 있지 않던 메서드라 제거했습니다!

minju412 committed:
feat: 스포티파이 사용자 정보를 조회하는 엔드포인트 추가

- 인증되지 않은 사용자가 보호된 uri에 접근할 경우 스포티파이 인증 페이지로 리다이렉트
- 스포티파이 로그인을 성공한 사용자는 원래 요청한 uri로 다시 리다이렉트
- 스포티파이 액세스토큰을 가져와 사용자 정보 요청

minju412 committed:
fix: Resolve Conflicts

minju412 committed:
Merge branch 'feature/issue#12' of https://github.com/petit-a-petit/MIML into feature/issue#12

minju412 committed:
comment: 불필요한 주석 제거

minju412 committed:
chore: 의존성 중복 제거

- gson 의존성 중복 제거

minju412 committed:
refacotr: provider와 providerId를 복합키로 사용

- provider와 providerId로 사용자를 식별함에 따라 기존 primary key였던 id가 식별자로서의 역할을 하지 않고 있음을 인지
- 이에 따라 provider와 providerId를 복합키로 사용하도록 변경

minju412 commented on build.gradle:
```diff
@@ -31,46 +30,21 @@ dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-mail:2.7.13'
 }
 
-jacocoTestCoverageVerification {
```
앗 jacoco 관련 코드는 건든 적이 없는 것 같은데 왜 제거되었는지 모르겠네요..🥲 다시 추가해 놓겠습니다!

minju412 committed:
chore: jacoco 설정 복구

- 의도치않게 누락된 코드 복구

minju412 committed:
refactor(CustomOAuth2User): isFirst 필드 제거

- oMember(=optionalMember)를 통해 처음 가입한 회원인지 판단이 가능하기 때문에 삭제
- 이미 가입한 회원인 경우 데이터베이스에 저장하지 않음

minju412 committed:
refactor(CustomOauth2UserService): 역할에 따른 메서드 분리

- 한 가지 메서드가 한 가지 역할만 하도록 메서드 분리
- 식별자를 구분하여 OAuth2UserInfo를 가져오는 역할, 기존 사용자를 가져오는 역할, 새로운 사용자를 생성하는 역할 등으로 메서드 분리

minju412 committed:
refactor: 불필요한 코드 제거

- 트랜잭션 관리가 필요하지 않은 부분이기 때문에 제거
